src/server/pkg/obj/testing/deployment_test.go

```diff
@@ -0,0 +1,359 @@
+package testing
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"testing"
+
+	"github.com/gogo/protobuf/proto"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kube "k8s.io/client-go/kubernetes"
+
+	"github.com/pachyderm/pachyderm/src/client"
+	"github.com/pachyderm/pachyderm/src/client/pfs"
+	"github.com/pachyderm/pachyderm/src/client/pkg/errors"
+	"github.com/pachyderm/pachyderm/src/client/pkg/require"
+	"github.com/pachyderm/pachyderm/src/client/pps"
+	"github.com/pachyderm/pachyderm/src/server/pkg/deploy/assets"
+	"github.com/pachyderm/pachyderm/src/server/pkg/serde"
+	tu "github.com/pachyderm/pachyderm/src/server/pkg/testutil"
+)
+
+// Change this to false to keep kubernetes namespaces around after the test for debugging purposes
+const cleanup = true
+
+// Rewrites kubernetes manifest services to auto-allocate external ports and
+// reduce cpu resource requests for parallel testing.
+type ManifestRewriter struct {
+	serde.Encoder
+}
+
+func rewriterCallback(innerCb func(map[string]interface{}) error) func(map[string]interface{}) error {
+	return func(data map[string]interface{}) error {
+		var err error
+		if innerCb != nil {
+			err = innerCb(data)
+		}
+		rewriteNodePort(data)
+		return err
+	}
+}
+
+func rewriteNodePort(data map[string]interface{}) {
+	if data["kind"] == "Service" {
```

Comment:
I'm still not 100% sure I understand these tests, but would it make sense to declare a const string JSON patch and then apply it? (and just make sure the patch was applied without errors)

I'll try to look up the library that does this, but we use it in PPS for `pod_patch`

Reply:
There are a couple hurdles with that approach.  First, the manifest isn't actually json, it's a set of newline-delimited json objects, so I'd have to parse and apply the patch to each separately (annoying but fine).  Second, the rewriting here is conditional, but JSON patches don't support that (I think) - we could either test the type of the object before applying (which means we need to parse the JSON), or just ignore errors (which may break silently if we aren't careful).  Finally, I don't see a way to write a patch that applies to all elements in a subarray (such as replacing the cpu for all containers in a deployment template), which would make the patch much harder to apply.

I gave a shot at moving this to jsonpatch (after writing half this comment) and I don't think it is worth it at the moment. 
 This code is a little fragile and ugly, but it's fairly straightforward and does the job for now.  If we ever have new requirements for this, we can reevaluate =(